### PR TITLE
Spiral Spanning Tree Coverage Path Planning

### PR DESCRIPTION
**Pull request by reso1**
Hi it's me again.
This is another CPP algorithm implementation,
proposed in [Spiral-STC: An On-Line Coverage Algorithm of Grid Environments
by a Mobile Robot](https://ieeexplore.ieee.org/abstract/document/1013479) by Gabriely et.al.

### CODE REVIEW

**Comment by AtsushiSakai** [comment]
Thank you again!!

File: PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+import sys
+import math
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+do_animation = True
+
+
+class SpiralSTC():
```
Comment (AtsushiSakai) [inline PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py:20]: redundant parentheses

File: PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+    def __init__(self, occmap):
```
Comment (AtsushiSakai) [inline PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py:21]: occmap -> occ_map

File: PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+            plt.show()
+
+
+if __name__ == "__main__":
+    dir_path = os.path.dirname(os.path.realpath(__file__))
+    img = plt.imread(os.path.join(dir_path, 'map', 'test_2.png'))
+    STC_planner = SpiralSTC(img)
+    start = (10, 0)
+    edge, route, path = STC_planner.plan(start)
+    STC_planner.viz_plan(edge, path, start)
```
Comment (AtsushiSakai) [inline PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py:303]: These code should be in `main()` function.

File: PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
+"""
+Terrain-Adaptive Spiral-STC Coverage Path Planner
```
Comment (AtsushiSakai) [inline PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py:2]: What `STC` stands for?

File: PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+        visit_times = np.zeros(
+            (self.merged_map_height, self.merged_map_width), dtype=np.int)
+        visit_times[start[0]][start[1]] = 1
+
+        route, path = [], []
+        # counter-clockwise neighbor finding order
+        order = [[1, 0], [0, 1], [-1, 0], [0, -1]]
+
+        def is_valid_node(i, j):
+            is_i_valid_bounded = i >= 0 and i < self.merged_map_height
```
Comment (AtsushiSakai) [inline PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py:54]: `0 <= i < self.merged_map_height` is more simpler.

File: PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+            is_j_valid_bounded = j >= 0 and j < self.merged_map_width
```
Comment (AtsushiSakai) [inline PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py:55]: `0 <= j < self.merged_map_width` is more simpler.

File: PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+            if is_i_valid_bounded and is_j_valid_bounded:
+                # free only when the 4 sub-cells are all free
+                return bool(
+                    self.occmap[2*i][2*j] and self.occmap[2*i+1][2*j]
+                    and self.occmap[2*i][2*j+1] and self.occmap[2*i+1][2*j+1])
+            return False
+
+        def STC(w, x):
```
Comment (AtsushiSakai) [inline PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py:63]: `w` is not used? and `x` should be `current_node`.

File: PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+                    for inc in order:
+                        ni, nj = node[0] + inc[0], node[1] + inc[1]
+                        if is_valid_node(ni, nj) and visit_times[ni][nj] == 0:
+                            has_node_with_unvisited_ngb = True
+                            break
+
+                    if has_node_with_unvisited_ngb:
+                        break
+
+        # generate route by recursing form start node
```
Comment (AtsushiSakai) [inline PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py:104]: typo? `recurring`?

File: PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+        for idx in range(len(route)-1):
+            dp = abs(route[idx][0] - route[idx+1][0]) + \
+                abs(route[idx][1] - route[idx+1][1])
+            if dp == 0:
+                # special handle for round-trip path
+                path.append(self.round_trip(route[idx-1], route[idx]))
+            elif dp == 1:
+                path.append(self.move(route[idx], route[idx+1]))
+            elif dp == 2:
+                # special handle for non-adjecent route nodes
```
Comment (AtsushiSakai) [inline PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py:117]: typo? 'adjacent'?

File: PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+                mid_node = self.get_intermediate_node(route[idx], route[idx+1])
+                path.append(self.move(route[idx], mid_node))
+                path.append(self.move(mid_node, route[idx+1]))
+            else:
+                sys.exit('adjecent path node distance larger than 2')
```
Comment (AtsushiSakai) [inline PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py:122]: type? 'adjacent'?

File: PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py [hunk 11]
```diff
@@ -1,5 +1,5 @@
 """
-Terrain-Adaptive Spiral-STC Coverage Path Planner
```
Comment (reso1) [inline PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py]: Corrected the method name

File: PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py [hunk 12]
```diff
@@ -17,18 +17,17 @@
...
+class SpiralSTC:
+    def __init__(self, occ_map):
+        self.origin_map_height = occ_map.shape[0]
+        self.origin_map_width = occ_map.shape[1]
 
-        # original map resolution must be 2^k
-        if math.log2(self.origin_map_height) % 1 == 0 and \
-                math.log2(self.origin_map_width) % 1 == 0:
-            sys.exit('original map width/height must be 2^k \
+        # original map resolution must be even
```
Comment (reso1) [inline PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py:25]: The old if condition is wrong, corrected it into the right one.

File: PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+        if len(itsc) == 0:
+            sys.exit('get_intermediate_node: \
+                 no intermediate node between', p, q)
+        elif len(itsc) == 1:
+            return list(itsc)[0]
+        else:
+            sys.exit('get_intermediate_node: \
+                more than 1 intermediate node between', p, q)
+
+    def viz_plan(self, edge, path, start):
```
Comment (AtsushiSakai) [inline PathPlanning/SpiralSpanningTreeCPP/spiral_spanning_tree_coverage_path_planner.py:228]: visualize_path?

**Comment by AtsushiSakai** [comment]
Thanks @reso1 . I will merge it.

**Pull request merged**